### PR TITLE
SDK: make it possible to use block builder script from other repositories 

### PR DESCRIPTION
**Pull request by simison**
We need a `bin` ([NPM docs](https://docs.npmjs.com/files/package.json#bin)) entry point in `package.json` to be able to use the builder from plugins like Jetpack.

Our `calypso-gutenberg-sdk` command points to `bin/sdk-cli.js` file which will in future hold logic for parsing arguments. In this PR that file is super crude and simple. It accepts only one task ("build-block") and simply passes arguments onwards to existing `bin/create-scripts/block.js`

The goal is that in Jetpack folder we should be able to run:
```bash
npx calypso-gutenberg-sdk build-block \
  block-file-relative-to-calypso.js \
  output-folder-relative-to-jetpack
```

As a new feature, block-builder script takes 3rd argument which is the absolute path of destination directory for the block. Defaulting to previous `./build` directory inside Calypso if none is provided.

Paths in scripts need to resolve to `./node_modules/wp-calypso`, not to Jetpack's root. I added [`context`](https://webpack.js.org/configuration/entry-context/#context) Webpack option to block-builder to help with that.

*The blocker* I'm having right now is that Babel doesn't seem to pick up presets. I get this error:

```bash
npx calypso-gutenberg-sdk build-block client/gutenberg/extensions/hello-dolly/hello-block.js $(pwd)/_inc/build


Hash: dd586a1798f6a5a83d17
Version: webpack 4.16.2
Time: 664ms
Built at: 07/24/2018 11:29:22 AM
                Asset    Size  Chunks  Chunk Names
blocks-hello-dolly.js  10 KiB       0  main
Entrypoint main = blocks-hello-dolly.js
[0] ./client/gutenberg/extensions/hello-dolly/hello-block.js 1.89 KiB {0} [built] [failed] [1 error]

ERROR in ./client/gutenberg/extensions/hello-dolly/hello-block.js
Module build failed (from ./node_modules/thread-loader/dist/cjs.js):
Thread Loader (Worker 0)
/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/client/gutenberg/extensions/hello-dolly/hello-block.js: Unexpected token (11:29)

   9 | 	icon: 'format-audio',
  10 | 	category: 'layout',
> 11 | 	edit: ( { isSelected } ) => <p>{ isSelected ? 'Editing Hello Dolly' : 'Viewing Hello Dolly' }</p>,
     | 	                            ^
  12 | 	save: () => <p>Saving Hello Dolly</p>,
  13 | } );
  14 |

    at Parser.raise (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:3898:15)
    at Parser.unexpected (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:5227:16)
    at Parser.parseExprAtom (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:6303:20)
    at Parser.parseExprSubscripts (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:5903:21)
    at Parser.parseMaybeUnary (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:5882:21)
    at Parser.parseExprOps (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:5791:21)
    at Parser.parseMaybeConditional (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:5763:21)
    at Parser.parseMaybeAssign (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:5710:21)
    at Parser.parseFunctionBody (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:6892:24)
    at Parser.parseArrowExpression (/Users/mikael/Documents/Work/Automattic/Calypso/wp-calypso/node_modules/@babel/parser/lib/index.js:6852:10)
```

I assumed Webpack's `context` would solve that but might be I need to use Babel's `sourceRoot` option somewhere as well. https://babeljs.io/docs/en/babel-core

### Next steps / todo
- We'll need to use [`yargs`](https://yargs.js.org/) or similar to stop passing "nth" arguments and use flags like `--output` instead. That logic should go to `sdk-cli.js`
- Block builder script should export a method so that we don't need to rely on calling it via command line. `sdk-cli.js` should be our only CLI-entry point going forward.
- Jetpack needs a script which operates `calypso-gutenberg-sdk` command. WIP PR: https://github.com/Automattic/jetpack/pull/9944

### Testing
Running this in Calypso folder should still work:
```bash
node bin/create-scripts/block.js client/gutenberg/extensions/hello-dolly/hello-block.js
```

For further testing, you need to link Calypso to Jetpack. Later on, we'll require Calypso via dependencies.

In Calypso folder:
```bash
npm link
```

In Jetpack folder:
```bash
npm link wp-calypso
```

You can now run this in Jetpack folder:
```bash
npx calypso-gutenberg-sdk build-block client/gutenberg/extensions/hello-dolly/hello-block.js $(pwd)/_inc/build
```

...you should see the babel error. To get past that, you could modify hello-dolly with:
```
edit: () => null,
save: () => null,
```

You should end up with a file `_inc/build/blocks-hello-dolly.js` in Jetpack.

### CODE REVIEW

File: bin/create-scripts/block.js [hunk 1]
```diff
@@ -1,18 +1,24 @@
+#!/usr/bin/env node
+
 /** @format */
 const path = require( 'path' );
 const webpack = require( 'webpack' );
 
+// TODO: move this to the npm script, see #26258
+process.env[ 'SKIP_FLAG_IMAGES' ] = 'true';
```
Comment (dmsnell) [inline bin/create-scripts/block.js]: can we leave this for a separate PR? the flags while annoying aren't getting in our way at least since we don't have to include them in the sites themselves.

(it's good to get rid of them but I don't see the need to in this PR)

Comment (simison) [reply bin/create-scripts/block.js]: Yep, I cherry-picked the commit from https://github.com/Automattic/wp-calypso/pull/26260 to make testing this PR easier for now. Once that's in, I'll rebase the commit away from this PR.

**Comment by tyxla** [comment]
#26260 has been merged; feel free to delete the cherry-picked commit and rebase to the latest master.

File: package.json [hunk 2]
```diff
@@ -9,6 +9,9 @@
     "url": "https://github.com/Automattic/wp-calypso.git"
   },
   "main": "index.js",
+  "bin": {
+    "wp-calypso-sdk": "./bin/create-scripts/block.js"
```
Comment (tyxla) [inline package.json]: We've been talking about using `calypso-gutenberg-sdk build-block` (see p4TIVU-93I-p2), but also `calypso-gutenberg-sdk` will suffice since that's the single script we have for now.

Comment (dmsnell) [reply package.json]: good catch on the name.

also, I wouldn't worry about trying to shorten it for now - we can be safe being verbose in the beginning and simplifying as time goes on.

File: package.json [hunk 3]
```diff
@@ -9,6 +9,9 @@
...
+  },
```
Comment (dmsnell) [inline package.json]: also, TIL how to make `bin` scripts in `npm` and it's much easier than I anticipated! thanks for that @simison

**Comment by simison** [comment]
I rebased, added a cli-wrapper, renamed the bin script and updated PR description.

Because of the `bin` script rename, you might need to run `npm link wp-calypso` in Jetpack folder again to update symlinks.

File: bin/sdk-cli.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+// First argument should always be the task name
+const task = args.shift();
+
+if ( ! args.length || task !== 'build-block' ) {
+	console.log( chalk.red( 'usage: npx calypso-gutenberg-sdk build-block <block>' ) );
+	process.exit(1);
+}
+
+execSync( `${ crossEnvShell } SKIP_FLAG_IMAGES=true node ${ buildBlockScript } ${ args.join( ' ' ) }`, {
```
Comment (simison) [inline bin/sdk-cli.js]: Now, I'd prefer to just pass our npm script (`npm run build-block ...`) directly to `execSync()` since that would be much cleaner but I didn't see any obvious way to do so: `execSync()` runs relative to `npx`, e.g. to Jetpack's root folder.

This whole file will get re-written in a follow-up PR very soon, so I don't mind having this a bit hacky at this point.

The main purpose of this file right now is just to solidify this API and unblock further cli-explorations: `npx calypso-gutenberg-sdk build-block ...`

Comment (simison) [reply bin/sdk-cli.js]: Eventually, we'll probably just do something like this instead:

```js
import { build } from 'sdk/block.js';

build( args );
```

File: package.json (hunk 2 above)
Comment (simison) [reply package.json]: Made it work as `calypso-gutenberg-sdk build-block ...` 🎉

File: bin/create-scripts/block.js [hunk 5]
```diff
@@ -1,18 +1,21 @@
...
 /** @format */
 const path = require( 'path' );
 const webpack = require( 'webpack' );
 
 const __rootDir = path.resolve( __dirname, '../../' );
-const entryPath = path.resolve( process.argv[ 2 ] );
+const entryPath = path.resolve( __rootDir, process.argv[ 2 ] );
 const sourceDir = path.dirname( entryPath );
-const outputDir = path.join( sourceDir, 'build' );
+const outputDir = process.argv[ 3 ] ? process.argv[ 3 ] : path.join( sourceDir, 'build' );
```
Comment (dmsnell) [inline bin/create-scripts/block.js]: fine for now, but after this let's add option-parsing so this doesn't get out of control with other necessary CLI options

Comment (simison) [reply bin/create-scripts/block.js]: Yep! 💯

File: bin/sdk-cli.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+const chalk = require( 'chalk' );
+const execSync = require( 'child_process' ).execSync;
+const path = require( 'path' );
+
+const buildBlockScript = path.resolve( __dirname, 'create-scripts/block.js' );
+const crossEnvShell = path.resolve( __dirname, '../node_modules/.bin/cross-env-shell' );
+let args = process.argv;
+
+// Remove `node` and SDK script
+args.splice(0, 2);
```
Comment (dmsnell) [inline bin/sdk-cli.js]: no need for the mutation here

```js
const [ /* node */, /* bin/create-scripts/block.js */, task, ...args ] = process.argv;
```

Comment (simison) [reply bin/sdk-cli.js]: Excellent :thumbsup: TIL

File: bin/sdk-cli.js (hunk 4 above)
Comment (dmsnell) [inline bin/sdk-cli.js]: we're kind of working around the system here - `exec` sets `ENV` key/value pairs for us

```js
execSync( `node ${buildBlockScript} ${ args.join( ' ' ) }`, {
	env: {
		SKIP_FLAG_IMAGES: true,
	},
	shell: true,
	stdio: 'inherit',
} )
```

at this point of passing args though why not use `spawn` instead?

```js
spawnSync( 'node', [ buildBlockScript, ...args ], {
	env: {
		...process.env, // is this needed?
		SKIP_FLAG_IMAGES: true,
	},
	shell: true,
	stdio: 'inherit',
} );
```

**Comment by simison** [comment]
Babel issue solved in https://github.com/Automattic/wp-calypso/pull/26300

**Comment by blowery** [comment]
> Later on, we'll require Calypso via dependencies.

@simison @dmsnell 

Whoa whoa whoa. So we're planning on publishing a semver compliant calypso?

**Comment by dmsnell** [comment]
hopefully @blowery we'll get to a point where we have pulled out Automattic's enhancements to Calypso and what will remain is a small core that's much easier to think about with regards to versioning.

that is, things like…
 - `components`
 - basic state system (without all our reducers)
 - basic data layer (without our handlers)
 - basic routing (long way off)
…will be included in the core while things like…
 - `my-sites`
 - `blocks`
 - reducers, selectors, handlers
…will attach in as extensions.

we actually have much less churn on what I think composes the real framework core than we do on features and if we get there we'll have a million and one benefits to reap from it, not the least of which is a decoupling of features from the framework which allows for significantly better caching and independence of pages and panels and features.
